[LexicalDestroyHoisting] Adopt iterative dataflow.

nate-chandler opened the pull request:
Adopt IterativeBackwardReachability and VisitBarrierAccessScopes.  Enables hoisting destroy_values over loops and avoiding hoisting destroy_values into unrelated access scopes, respectively.

rdar://93369506

nate-chandler committed:
[LexicalDestroyHoisting] Adopt iterative dataflow.

Adopt IterativeBackwardReachability.  Enables hoisting destroy_values
over loops.

rdar://93369506

nate-chandler commented:
@swift-ci please smoke test

nate-chandler committed:
[Gardening] Recapitalized type.

nate-chandler commented:
@swift-ci please test

atrick commented on include/swift/SILOptimizer/Analysis/VisitBarrierAccessScopes.h:
```diff
@@ -0,0 +1,190 @@
+  /// * These are backedges in the reversed graph, that are ignored in a reverse
+  /// depth-first search.
+  void visit() {
+    visitSortedTopologicallyBackwards(function, visitor.roots(), *this);
```
The casual reader should just see "visit blocks in forward order" here. But it takes a lot of mental effort to figure that out. Renaming it to `visitForwardTopological` would help. potentially as a separate PR if you want

nate-chandler replied:
Will fix in https://github.com/apple/swift/pull/58791.

atrick commented:
Same comment for LexicalDestroyHoisting...

The boilerplate detracts from readability. But I think that's mostly solved by moving the small method bodies in-line with the class definition--the ones that just provide abstraction or implement template interfaces, but don't have any nontrivial switches or loops.

nate-chandler committed:
[LexicalDestroyHoisting] Adopt VisitBarrierAccessScopes.

Avoids hoisting destroy_values into unrelated access scopes which could
introduce exclusivity violations.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please test windows platform

atrick commented on lib/SILOptimizer/Utils/LexicalDestroyHoisting.cpp:
```diff
@@ -193,26 +189,15 @@ bool Dataflow::classificationIsBarrier(Classification classification) {
-
-bool Dataflow::checkReachableBarrier(SILInstruction *instruction) {
+Dataflow::Effect Dataflow::effectForInstruction(SILInstruction *instruction) {
+  if (llvm::find(uses.ends, instruction) != uses.ends.end())
```
It looks like calling `effectForInstruction` for each instruction is now quadratic.

nate-chandler replied:
Fixed this issue in LexicalDestroyHoisting, ShrinkBorrowScope, and SSADestroyHoisting: https://github.com/apple/swift/pull/59057
